### ISSUE TITLE
Add tests for command-line and Python bindings

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1152]
This is a great issue for anyone looking to get involved with mlpack development.  No huge amount of C++ or machine learning knowledge is necessary; in order to write a test, essentially all you need to know is how the particular binding you are writing a test for works, and you can use the existing test code as boilerplate for the new tests.  So I think this is a good place to get started.

If you don't know about the automatic bindings system, you can read about it at http://www.mlpack.org/docs/mlpack-git/doxygen/bindings.html .

In essence, the problem is that we have all of these command-line programs and Python bindings like `mlpack_knn`, `mlpack_sparse_coding`, and so forth, but these are not rigorously tested.  Thanks to the automatic binding system, we now have an easy way to test these, which allows us to avoid all kinds of little bugs that have happened in the past where a command-line program might access the name of a variable incorrectly or something like this.  However, the tests have not been written yet.  So the purpose of this issue is to ensure that these tests actually do get written.

So, for example, let's take a look at a test we already have for the PCA binding.  This is from `src/mlpack/tests/pca_test.cpp`.

```
/**
 * Check that we can't specify an invalid new dimensionality.
 */
BOOST_AUTO_TEST_CASE(PCATooHighNewDimensionalityTest)
{ 
  arma::mat x = arma::randu<arma::mat>(5, 5);
  
  SetInputParam("input", std::move(x));
  SetInputParam("new_dimensionality", (int) 7); // Invalid.

  Log::Fatal.ignoreInput = true;
  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
  Log::Fatal.ignoreInput = false;
}
```

This is a pretty simple test, which just checks that a user can't ask for a new dimensionality that is too high, and this is the kind of test we're looking for here.  We don't need to check if the exact output of the algorithm is correct---we are already doing that in the other tests we have for the algorithms.  Instead we want to test the input/output of these bindings.

So, if you'd like to write a test, the basic procedure is this:

 * Pick an uncompleted binding from the list below (I am trying to keep this up to date):
   * [x] adaboost (`src/mlpack/methods/adaboost_main.cpp`): AdaBoost
   * [x] approx_kfn (`src/mlpack/methods/approx_kfn_main.cpp`): approximate k-nearest-neighbor search
   * [x] cf (`src/mlpack/methods/cf/cf_main.cpp`): collaborative filtering
   * [x] dbscan (`src/mlpack/methods/dbscan/dbscan_main.cpp`): DBSCAN clustering
   * [x] decision_tree (`src/mlpack/methods/decision_tree/decision_tree_main.cpp`): decision trees
   * [x] det (`src/mlpack/methods/det/det_main.cpp`): density estimation trees
   * [x] emst (`src/mlpack/methods/emst/emst_main.cpp`): Euclidean minimum spanning tree calculation
   * [x] fastmks (`src/mlpack/methods/fastmks/fastmks_main.cpp`): fast max-kernel search (#1356, needs someone to finish the work; you can use that PR as a starting point)
   * [x] gmm_generate (`src/mlpack/methods/gmm/gmm_generate_main.cpp`): generate samples from a GMM
   * [x] gmm_probability (`src/mlpack/methods/gmm/gmm_probability_main.cpp`): estimate the probability of a sample coming from a GMM
   * [x] gmm_train (`src/mlpack/methods/gmm/gmm_train_main.cpp`): train a GMM on data
   * [x] hmm_generate (`src/mlpack/methods/hmm/hmm_generate_main.cpp`): generate samples from an HMM
   * [x] hmm_loglik (`src/mlpack/methods/hmm/hmm_loglik_main.cpp`): find the log-likelihood of a sequence coming from an HMM
   * [x] hmm_train (`src/mlpack/methods/hmm/hmm_train_main.cpp`): train an HMM on data
   * [x] hmm_viterbi (`src/mlpack/methods/hmm/hmm_viterbi_main.cpp`): predict the hidden state sequence with an HMM
   * [x] hoeffding_tree (`src/mlpack/methods/hoeffding_trees/hoeffding_tree_main.cpp`): streaming decision trees
   * [x] kernel_pca (`src/mlpack/methods/kernel_pca/kernel_pca_main.cpp`): kernel PCA (#1600)
   * [x] kfn (`src/mlpack/methods/neighbor_search/kfn_main.cpp`): k-furthest neighbors search (#1263)
   * [x] kmeans (`src/mlpack/methods/kmeans/kmeans_main.cpp`): k-means clustering
   * [x] knn (`src/mlpack/methods/neighbor_search/knn_main.cpp`): k-nearest neighbors search (#1263)
   * [x] linear_regression (`src/mlpack/methods/linear_regression/linear_regression_main.cpp`): linear regression
   * [x] local_coordinate_coding (`src/mlpack/methods/local_coordinate_coding/lcc_main.cpp`): local coordinate coding
   * [x] logistic_regression (`src/mlpack/methods/logistic_regression/logistic_regression_main.cpp`): logistic regression
   * [x] lsh (`src/mlpack/methods/lsh/lsh_main.cpp`): locality-sensitive hashing
   * [x] mean_shift (`src/mlpack/methods/mean_shift/mean_shift_main.cpp`): mean shift clustering
   * [x] nbc (`src/mlpack/methods/naive_bayes/nbc_main.cpp`): naive Bayes classifier
   * [x] nca (`src/mlpack/methods/nca/nca_main.cpp`): neighborhood components analysis (#1369)
   * [x] nmf (`src/mlpack/methods/nmf/nmf_main.cpp`): non-negative matrix factorization
   * [x] pca (`src/mlpack/methods/pca/pca_main.cpp`): principal components analysis
   * [x] perceptron (`src/mlpack/methods/perceptron/perceptron_main.cpp`): perceptrons
   * [x] preprocess_binarize (`src/mlpack/methods/preprocess/preprocess_binarize_main.cpp`): binarize data
   * [x] preprocess_describe (`src/mlpack/methods/preprocess/preprocess_describe_main.cpp`): describe a dataset (#1337)
   * [x] preprocess_imputer (`src/mlpack/methods/preprocess/preprocess_imputer_main.cpp`): impute values into a dataset
   * [x] preprocess_split (`src/mlpack/methods/preprocess/preprocess_split_main.cpp`): split a dataset into a training and test set
   * [x] radical (`src/mlpack/methods/radical/radical_main.cpp`): an independent components analysis technique
   * [x] random_forest (`src/mlpack/methods/random_forest/random_forest_main.cpp`): random forests for classification
   * [x] range_search (`src/mlpack/methods/range_search/range_search_main.cpp`): range search
   * [x] softmax_regression (`src/mlpack/methods/softmax_regression/softmax_regression_main.cpp`): softmax regression
   * [x] sparse_coding (`src/mlpack/methods/sparse_coding/sparse_coding_main.cpp`): sparse coding

 * Use and understand the options for the binding that you have chosen, either from the command-line or from Python.

 * Based on some of the other examples in `src/mlpack/tests/main_tests/`, add a new file that tests the binding you chose.  Focus on tests like:
   * Ensuring errors are given when the user inputs invalid parameters.
   * Make sure that saved models can be reused in future invocations of the binding.
   * Make sure that output matrices have the expected shape.
   * *Don't* write tests to check the exact correctness of the algorithm---we already have those, so there is no need. :)

 * When you have good coverage of the binding, go ahead and open a PR and we can review it and get it merged.

In addition, *there's no need to ask whether or not you can work on a particular test*.  If no PR is open for the given binding and there is no test already in the repository, you are free to work on it and open a PR for it.  I'll try to keep the list above up to date.

**Comment by wkk91193** [comment]
@rcurtin  I'd like to take this...

**Comment by rcurtin** [comment]
@wkk91193: go ahead---pick a test and go for it.  From the description:

> In addition, there's no need to ask whether or not you can work on a particular test. If no PR is open for the given binding and there is no test already in the repository, you are free to work on it and open a PR for it.

**Comment by wkk91193** [comment]
@rcurtin I tried running the sample test fixture in src/mlpack/tests/main/pca_test.cpp on Clion, but I face [this](https://pastebin.com/aT5vpFav). Am I missing out somthing here?

**Comment by rcurtin** [comment]
I see you posted to the mailing list also, so I'll answer there.

**Comment by gurupunskill** [comment]
I'd like pick a few tests too.

**Comment by rcurtin** [comment]
Sure, go ahead, looks like a PR is open for EMST (#1165) now.

**Comment by evgfreyman** [comment]
I was looking to get involved into this project and tried to write a couple of tests, but met a problem: if you follow proposed way of writing tests (similar to src/mlpack/tests/main_tests/pca_test.cpp) then you get mlpackMain and mlpack::bindings::tests::programName redefined. Although it's relatively easy to handle the first issue, the second is much harder. The problem is that different programs define different parameters that inside constructors rely on global variable - programName, which meant to be different for different programs. I couldn't think up any fast workaround.

To my view, the proper solution is to explicitly define unique name of each test suite (only for binding tests) and pass it to parameters declaration. There is a chance that through macros' it can be done without modifying bindings itself. What do you think? Are there better solutions? I'd like to tackle this issue if you don't mind

**Comment by evgfreyman** [comment]
Just noticed PR for EMST (#1165) and it looks that it was built. Do I misunderstand something?

**Comment by rcurtin** [comment]
@evgfreyman: nice catch, you are right that this is an issue.  I had only made one test binding and mistakenly assumed the rest would work much in the same way that the Python bindings work.  However in this case, each binding is all put into _one_ .so instead of with the Python bindings where they all go into individual .sos, and the two issues you pointed out wouldn't be a problem.

I can see the difficulty of both problems, but I don't see any easy route to a fix on the `mlpackMain()` issue.  Can you elaborate on what you're thinking?  For the `programName` issue there is some extra complexity.

I think that the right solution to the `programName` problem is to entirely remove `programName` and add an extra parameter to the `TestOption<>` constructor that has the program name.  Basically the reason for `programName` is so that we can have the `CLI` module (which stores the command-line options) store different _sets_ of options that we can restore or remove at will.  This turns out to be needed for the Python bindings.

Let me think a little more about the `programName` problem.  The solution I described above requires modification to the `PARAM_*` macros, and I am not sure I like that, so I need to think if that can be avoided.

Thanks again for pointing this out.  As a last note, #1165 doesn't fail because the new `emst_test.cpp` file actually isn't being compiled.

**Comment by evgfreyman** [comment]
As for `mlpackMain()` the simplest way is to rename it to be program specific, for example  `mlpackMainPca()` and add following to the end of program:

```
#if not (BINDING_TYPE == BINDING_TYPE_TEST)
  void mlpackMain()
  {
    mlpackMainPca();
  }
#endif

```

However in binding tests we will need to access mlpackMainPca() directly.

Another approach is to have unique definition at each binding test, for example `#define PROGRAM_NAME pcaProgramName` and define following macros (for example at `mlpack_main.cpp`):

```
#ifndef PROGRAM_NAME
#define PROGRAM_NAME
#endif
#define TOKENPASTE(x, y) x ## y
#define TOKENPASTE2(x, y) TOKENPASTE(x, y)
#define MAIN TOKENPASTE2(mlpackMain, PROGRAM_NAME)
```

After this we can use the same naming `MAIN()` in both binding tests and programs itself

As for ways of handling `programName` let me think a bit...

**Comment by evgfreyman** [comment]
I manage to handle both issues . I added PR #1172 I guess we can continue discussion there

**Comment by rcurtin** [comment]
PR #1172 merged (thanks again @evgfreyman) so the problems mentioned are resolved and the code is updated to handle new binding tests.

**Comment by coffe2code** [comment]
I'd like to take this on.

**Comment by manish7294** [comment]
Opened PR #1193 for random_forest_main.cpp, nbc_main.cpp, decision_stump_main.cpp, perceptron_main.cpp & softmax_regression_main.cpp. @rcurtin please mark the tests which are done.

**Comment by KARTHEEKCIC** [comment]
I would like to work on binding tests for logistic regression. If any one is working, please reply.

**Comment by rcurtin** [comment]
It seems like I was incorrectly updating the checkboxes before, sorry about that.  Hopefully the changes "stick" now...

**Comment by manish7294** [comment]
I will be working next on sparse_coding, det, cf, kmeans and neighbor_search(kfn & knn). If any one is currently working on any of them then please notify.

**Comment by nikhilgoel1997** [comment]
I am working on binding tests for adaboost. If anyone else is working on this, please tell.

**Comment by Prabhat-IIT** [comment]
I am working on it @nikhilgoel1997

**Comment by prashanthduvvada** [comment]
Hello,

I'm newbie to mlpack. I would like to work on hoeffding_tree, kernel_pca, knn and nca. Since being new, I will take some time in doing this task but will send PR's as soon as possible. If anyone else is working on this please update.

**Comment by rcurtin** [comment]
Hi Prashanth; sorry but hoeffding_tree is already taken by #1211.  I think the other three are not taken though.  I would suggest only starting with one though.

**Comment by prashanthduvvada** [comment]
Oops! Didn't check the #1211. Then I would start with kernel_pca.

**Comment by akhandait** [comment]
I am working on range_search binding tests.

**Comment by Wenhao-H** [comment]
Hi I am working on nmf. If anyone is working on this pls let me know.

**Comment by rcurtin** [comment]
I am trying to keep the list of "taken" methods up to date, so if you see that I haven't put a check mark next to something you are working on, just let me know. :)

**Comment by akhandait** [comment]
I am having some issue with building mlpack_test with range_search_test.cpp. I have posted it on irc channel. Till it's solved, I am working on knn_test.

**Comment by robertohueso** [comment]
I am working on knn tests :)

**Comment by akhandait** [comment]
@robertohueso, I have already written 7 tests for knn_main and working on more, can you please chose some other method to write tests for?

**Comment by akhandait** [comment]
As test requirements for kfn and knn are almost all the same, I will commit tests for kfn as well in #1263.

**Comment by Wenhao-H** [comment]
@manish7294 hi manish have you started with cf binding test? cuz there there is a small issue with cf binding which I want to fix before you start. Or if you want I can help with implementing the cf binding test :)

**Comment by manish7294** [comment]
@Wenhao-H Currently I am going through my exams. So,  you can start with cf bindings if you like to. If you have any other questions then feel free to ask, I will try my best to reply as early as I can :)

**Comment by KARTHEEKCIC** [comment]
I would like to work on gmm_train, gmm_generate, gmm_probability. If anyone is working please let me know.

**Comment by yamidark** [comment]
Hi, I'm new to mlpack, and I would like to work on mean_shift.

**Comment by rcurtin** [comment]
Hi @yamidark, I've marked it as taken for you---open a PR when you are ready and we can review it.

**Comment by Namrata96** [comment]
@rcurtin Hi, I'm new to mlpack, and would like to work on approx_kfn.

**Comment by rcurtin** [comment]
Hi @Namrata96, of course, feel free.  I'll mark it as taken.

**Comment by manish7294** [comment]
@rcurtin Finally I got over with my academics :)
I am thinking of finishing up this part, It shouldn't take much time. Is the above three methods still available?

**Comment by rcurtin** [comment]
@manish7294: yeah, there have been no PRs opened for those.  I see you opened #1384, so I'll review that shortly. :+1:

**Comment by sourabhvarshney111** [comment]
@rcurtin @zoq Looks like det tests are also complete and hence the issue can be closed.

**Comment by manish7294** [comment]
@sourabhvarshney111 I think it will be way too early to close this issue. Still, there are many tests which are marked but not done or have open pull requests. So, I think it will be best to leave a note mentioning them in case someone wishes to take them up.

**Comment by sourabhvarshney111** [comment]
Thanks for clarifying @manish7294. Let me check which are open and still are not complete.

**Comment by manish7294** [comment]
# Here is a list:
**Still Available:**

- gmm_generate
- gmm_probability
- gmm_train
- local_coordinate_coding
- range_search
 

Open PR:

- preprocess_describe #1337
- fastmks #1356 
- kfn #1263 

Maybe a few more are there :)

**Comment by KARTHEEKCIC** [comment]
I will open PR for gmm_generate, gmm_probability, gmm_train soon.

**Comment by rcurtin** [comment]
Thanks everyone for the clarifications.  I updated the list at the top.  The issue is almost fully resolved now. :+1:

**Comment by gauravcr7rm** [comment]
Hi, I'm new to mlpack, and would like to work on gmm_generate, gmm_probability and gmm_train.

**Comment by rcurtin** [comment]
@guaravcr7rm @niteya-shah sounds good, looking forward to the results. :+1:

**Comment by bhavya01** [comment]
Hi, I would like to work on local coordinate coding.

**Comment by rcurtin** [comment]
@bhavya01 please go ahead, I look forward to a PR. :+1:

**Comment by rcsunanda** [comment]
Hello, I'm going to work on the fastmks test (fast max-kernel search). Sunanda.

**Comment by Yashwants19** [comment]
Hello @rcurtin, I'm would like to work on the fastmks test for command-line and Python bindings . If no one is working on this

**Comment by rcsunanda** [comment]
Hi, @Yashwants19 , I was meaning to work on the fastmks test, but I haven't been able to get started yet. So yes, please go ahead with it.

**Comment by rcurtin** [comment]
Hey @Yashwants19 and @rcsunanda, there is actually an old PR you can use as a starting point for the FastMKS tests: #1356. If I remember right there were still a few issues that needed to be debugged for that one.

**Comment by Hsankesara** [comment]
Hey @rcurtin, I like to work to on KNN command line and Python binding.

**Comment by rcurtin** [comment]
Hi @Hsankesara, sounds good.  There is an old PR for that too that you could look at as a starting point: #1263. :+1:

**Comment by rcurtin** [comment]
Oops, forgot to mark it 'keep open'.

**Comment by rcurtin** [comment]
I can't believe it, all of the tests are finally added.  Thank you everyone for your hard work and contributions. :+1:  I'll go ahead and close the issue now. :)
